conman/routes/tests/test_models.py

```diff
@@ -60,10 +62,41 @@ def test_create_leaf_without_slug(self):
         root_route = RootRouteFactory.create()
         leaf = RouteFactory.build(slug='', parent=root_route)
 
-        with self.assertRaises(ValueError):
+        with self.assertRaises(ValidationError):
             leaf.save()
 
 
+class RouteFormValidate(TestCase):
+    """Check validation of Route slugs and ancestry applies to forms."""
+    def setUp(self):
+        """Create a simple Route form."""
+        class RouteForm(ModelForm):
+            class Meta:
+                model = Route
+                exclude = []
+
+        self.form = RouteForm
+
+    def test_validate_root_with_slug(self):
+        """Root must not have a slug."""
+        data = {'slug': 'slug', 'parent': None}
+
+        form = self.form(data)
+        form.is_valid()
+
+        self.assertIn('__all__', form.errors)
+
+    def test_validate_leaf_without_slug(self):
+        """Leaf Routes must have a slug."""
+        root_route = RootRouteFactory.create()
+        data = {'slug': '', 'parent': root_route.pk}
+
+        form = self.form(data)
+        form.is_valid()
+
+        self.assertIn('__all__', form.errors)
```

Comment:
This `form.errors` is:

```python
{'slug': ['This field is required.'], '__all__': ['Route can be a root, or have a slug, not both.']}
```

We can remove the `slug` error by adding `blank=True` to `Route.slug`, but I'm not sure if we need to.